dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/cache/TimeToLive.java

```diff
@@ -0,0 +1,154 @@
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.hisp.dhis.analytics.cache;
+
+import static java.time.LocalDateTime.now;
+import static java.time.LocalDateTime.ofInstant;
+import static java.time.ZoneOffset.UTC;
+import static java.time.temporal.ChronoUnit.DAYS;
+import static java.time.temporal.ChronoUnit.YEARS;
+import static org.hisp.dhis.analytics.cache.TimeToLive.Periods.MONTHLY;
+import static org.hisp.dhis.analytics.cache.TimeToLive.Periods.QUARTERLY;
+import static org.hisp.dhis.analytics.cache.TimeToLive.Periods.SIX_MONTHS;
+import static org.hisp.dhis.analytics.cache.TimeToLive.Periods.WEEKLY;
+import static org.hisp.dhis.setting.SettingKey.ANALYTICS_CACHE_TIMEOUT_FACTOR_MONTHLY_PERIOD_IN_SECONDS;
+import static org.hisp.dhis.setting.SettingKey.ANALYTICS_CACHE_TIMEOUT_FACTOR_QUARTERLY_PERIOD_IN_SECONDS;
+import static org.hisp.dhis.setting.SettingKey.ANALYTICS_CACHE_TIMEOUT_FACTOR_SIX_MONTHS_PERIOD_IN_SECONDS;
+import static org.hisp.dhis.setting.SettingKey.ANALYTICS_CACHE_TIMEOUT_FACTOR_WEEKLY_PERIOD_IN_SECONDS;
+import static org.hisp.dhis.setting.SettingKey.ANALYTICS_CACHE_TIMEOUT_FACTOR_YEARLY_OR_OVER_PERIOD_IN_SECONDS;
+import static org.springframework.util.Assert.notNull;
+
+import java.time.Instant;
+
+import org.hisp.dhis.analytics.DataQueryParams;
+import org.hisp.dhis.setting.SettingKey;
+import org.hisp.dhis.setting.SystemSettingManager;
+
+public class TimeToLive
+    implements
+    Computable
+{
+
+    static final long DEFAULT_TO_30_SECONDS = 30;
+
+    private final DataQueryParams params;
+
+    private SystemSettingManager systemSettingManager;
+
+    public TimeToLive( final DataQueryParams params, final SystemSettingManager systemSettingManager )
+    {
+        notNull( params, "Object params must not be null" );
+        notNull( params.getEarliestStartDate(), "Object params.getEarliestStartDate() must not be null" );
+        notNull( params.getLatestEndDate(), "Object params.getLatestEndDate() must not be null" );
+        notNull( systemSettingManager, "Object systemSettingManager must not be null" );
+        this.params = params;
+        this.systemSettingManager = systemSettingManager;
+    }
+
+    /**
+     * Calculates the time to live based on specific internal rules.
+     *
+     * @return the computed TTL value in SECONDS.
+     */
+    @Override
+    public long compute()
+    {
+        final Instant oldestDate = params.getEarliestStartDate().toInstant();
+        final Instant mostRecentDate = params.getLatestEndDate().toInstant();
+        final long daysDiff = DAYS.between( oldestDate, mostRecentDate );
+        return calculationFor( daysDiff, mostRecentDate );
+    }
+
+    /**
+     * Execute the internal rules in order to calculate a TTL for the given
```

Comment:
Make the javadoc describe what is actually happening.

Reply:
done